src/emqx_protocol.erl

```diff
@@ -421,7 +423,96 @@ process_packet(?DISCONNECT_PACKET(?RC_SUCCESS, #{'Session-Expiry-Interval' := In
 process_packet(?DISCONNECT_PACKET(?RC_SUCCESS), PState) ->
     {stop, normal, PState#pstate{will_msg = undefined}};
 process_packet(?DISCONNECT_PACKET(_), PState) ->
-    {stop, normal, PState}.
+    {stop, normal, PState};
+
+process_packet(?AUTH_PACKET(_), PState = #pstate{auth_state = disconnected}) ->
+    {error, proto_not_connected, PState};
+process_packet(?AUTH_PACKET(_), PState = #pstate{auth_method = <<>>, auth_state = AuthState}) ->
+    if 
+        AuthState =:= authenticating ->
+            connack({?RC_PROTOCOL_ERROR, PState});
+        true ->     % reauthenticating or connected 
+            disconnect({?RC_PROTOCOL_ERROR, PState})
+    end;
+process_packet(?AUTH_PACKET(_, undefined), PState = #pstate{auth_state = AuthState}) ->
+    if 
+        AuthState =:= authenticating ->
+            connack({?RC_MALFORMED_PACKET, PState});
+        true ->     % reauthenticating or connected 
+            disconnect({?RC_MALFORMED_PACKET, PState})
+    end;
+process_packet(?AUTH_PACKET(?RC_CONTINUE_AUTHENTICATE, _), PState = #pstate{auth_state = connected}) -> 
+    disconnect({?RC_PROTOCOL_ERROR, PState});
+process_packet(?AUTH_PACKET(?RC_CONTINUE_AUTHENTICATE, Properties), PState) ->
+    process_enhanced_auth(maps:get('Authentication-Method', Properties, undefined), maps:get('Authentication-Data', Properties, undefined), PState);
+process_packet(?AUTH_PACKET(?RC_RE_AUTHENTICATE, Properties), PState = #pstate{auth_state = connected}) ->
+    process_enhanced_auth(maps:get('Authentication-Method', Properties, undefined), maps:get('Authentication-Data', Properties, undefined), PState);
+process_packet(?AUTH_PACKET(?RC_RE_AUTHENTICATE), PState = #pstate{auth_state = reauthenticating}) ->
+    disconnect({?RC_PROTOCOL_ERROR, PState});
+process_packet(?AUTH_PACKET(?RC_RE_AUTHENTICATE), PState = #pstate{auth_state = authenticating}) ->
+    connack({?RC_PROTOCOL_ERROR, PState}).
+
+%%------------------------------------------------------------------------------
+%% process enhanced authenticate
+%%------------------------------------------------------------------------------
+
+process_enhanced_auth(AuthMethod, _AuthData, PState = #pstate{auth_state = AuthState}) 
+    when AuthMethod =:= undefined orelse AuthMethod =:= <<>> ->
+    if 
+        AuthState =:= reauthenticating orelse AuthState =:= connected ->
+            disconnect({?RC_MALFORMED_PACKET, PState});
+        true ->
+            connack({?RC_MALFORMED_PACKET, PState})
+    end;
+process_enhanced_auth(AuthMethod, AuthData, PState = #pstate{auth_method = OldAuthMethod, auth_state = AuthState}) ->
+    if 
+        OldAuthMethod =/= <<>> andalso OldAuthMethod =/= AuthMethod ->
```

Comment:
this clause should be delegated to a function or variable with a more readable name.
e.g. `is_bad_auth_method(OldAuthMethod, AuthMethod)`